pallets/crowdloans/src/lib.rs

```diff
@@ -959,6 +979,115 @@ pub mod pallet {
+            let mut all_refunded = false;
+
+            let vault = Self::current_vault(crowdloan).ok_or(Error::<T>::VaultDoesNotExist)?;
+            
+            // 1. check phase, should be Closed or Failed
+            ensure!(
+                vault.phase == VaultPhase::Closed || vault.phase == VaultPhase::Failed,
+                Error::<T>::IncorrectVaultPhase
+            );
+
+            for child_storage_kind in [
+                ChildStorageKind::Contributed,
+                ChildStorageKind::Flying,
+                ChildStorageKind::Pending,
+            ] {
+                let contribution =
+                    Self::contribution_iterator(vault.trie_index, child_storage_kind);
+
+                for (who, (amount, _referral_code)) in contribution {
+                    if refund_count >= T::RemoveKeysLimit::get() {
+                        all_refunded = true;
+                        return Ok(());
+                    }
+                    refund_count += 1;
+
+                    // Mint relay currency
+                    T::Assets::mint_into(vault.trie_index, &who, amount)?;
+                    Self::contribution_kill(vault.trie_index, &who, child_storage_kind);
+                }
+            }
+
+            // If all contributions have been refunded then return AllRefunded event
+            // if not then return PartiallyRefunded event
+            if all_refunded {
+                Self::deposit_event(Event::<T>::AllRefunded(crowdloan, (lease_start, lease_end)));
+            } else {
+                Self::deposit_event(Event::<T>::PartiallyRefunded(
+                    crowdloan,
+                    (lease_start, lease_end),
+                ));
+            }
+
+            Ok(())
+        }
+
+        #[pallet::weight(<T as Config>::WeightInfo::dissolve_vault())]
+        #[transactional]
+        pub fn dissolve_vault(
+            origin: OriginFor<T>,
+            crowdloan: ParaId,
+            lease_start: LeasePeriod,
+            lease_end: LeasePeriod,
+        ) -> DispatchResult {
+            // users who can create vaults can dissolve them
+            T::DissolveVaultOrigin::ensure_origin(origin)?;
+
+            // 1. check phase, should be Closed or Failed or Expired
+            let mut vault = Self::vaults((&crowdloan, &lease_start, &lease_end))
+                .ok_or(Error::<T>::VaultDoesNotExist)?;
+
+            ensure!(
+                vault.phase == VaultPhase::Closed
+                    || vault.phase == VaultPhase::Failed
+                    || vault.phase == VaultPhase::Expired,
+                Error::<T>::IncorrectVaultPhase
+            );
+
+            let has_contributed = Self::has_childstorage(&mut vault);
+
+            ensure!(!has_contributed, Error::<T>::NotReadyToDissolve);
+
+            ensure!(
+                vault
+                    .contributed
+                    .checked_add(Self::total_contribution(&mut vault, 0)?)
```

Comment:
total_contribution should be:

```
vault.pending.checked_add(vault.flying).and_then(|sum| sum.checked_add(vault.contributed))
```

Reply:
@GopherJ  In here I'm using the existing function by passing `0` as the amount. It does the same thing . Did this as per the request on :- https://github.com/parallel-finance/parallel/pull/1168#discussion_r786469790

```rust
        fn total_contribution(
            vault: &Vault<T>,
            amount: BalanceOf<T>,
        ) -> Result<BalanceOf<T>, ArithmeticError> {
            vault
                .contributed
                .checked_add(vault.flying)
                .and_then(|sum| sum.checked_add(vault.pending))
                .and_then(|sum| sum.checked_add(amount))
                .ok_or(ArithmeticError::Overflow)
        }
```